wallet/keypass-json.js

```diff
@@ -0,0 +1,150 @@
+const scrypt = require('scryptsy')
+const cbor = require('cbor')
+const chacha20 = require('@stablelib/chacha')
+
+const {pbkdf2Async} = require('./helpers/pbkdf2')
+const hashBlake2b256 = require('./helpers/hashBlake2b256')
+const {HARDENED_THRESHOLD} = require('./constants')
+
+function transformPassword(password) {
+  return password ? hashBlake2b256(password, null, 32) : Buffer.from([])
+}
+
+function verifyPassword(passwordToVerify, passwordHash) {
+  const passwordHashSplit = cbor
+    .decode(passwordHash)
+    .toString('ascii')
+    .split('|')
+
+  const n = 1 << parseInt(passwordHashSplit[0], 10)
+  const r = parseInt(passwordHashSplit[1], 10)
+  const p = parseInt(passwordHashSplit[2], 10)
+  const salt = Buffer.from(passwordHashSplit[3], 'base64')
+  const expectedHash = Buffer.from(passwordHashSplit[4], 'base64')
+
+  const passwordToVerifyHash = scrypt(
+    cbor.encode(transformPassword(passwordToVerify)),
+    salt,
+    n,
+    r,
+    p,
+    expectedHash.length
+  )
+
+  return passwordToVerifyHash.equals(expectedHash)
+}
+
+function getRandomSaltForPasswordHash() {
+  const randBytes = Buffer.alloc(32)
+  window.crypto.getRandomValues(randBytes)
+
+  return cbor.encode(randBytes)
+}
+
+function hashPasswordAndPack(password, salt) {
+  const [n, r, p, hashLen] = [14, 8, 1, 32]
+  const hash = scrypt(cbor.encode(transformPassword(password)), salt, 1 << n, r, p, hashLen)
+
+  return [
+    n.toString(),
+    r.toString(),
+    p.toString(),
+    salt.toString('base64'),
+    hash.toString('base64'),
+  ].join('|')
+}
+
+async function encryptOrDecryptWalletSecret(walletSecret, password) {
```

Comment:
This is ambiguous and also in my opinion breaks single responsibility principle. I suggest to make two functions instead of this one.

Reply:
The purpose of the name of the function is to make clear that the function is self inverse, but I split it anyway, putting the crypto stuff in the "memoryCombine" function - similarly to the source code in cardano-crypto. I tried making two separate functions but I didn't like it, because at first sight you may not understand if the functions decrypt and encrypt do the same thing, which is, I believe quite an important insight.

Reply:
One "hacky" approach I tried is the following:
const [encryptWalletSecret, decryptWalletSecret] = Array(2).fill(async(x,y) => {...})

But I think this goes quite a bit against readability, so I dropped it, but maybe you think otherwise.